### ISSUE TITLE
Polyfon fill with NaNs cause excursion to origin

### ISSUE THREAD

**Issue by joa-quim** [GenericMappingTools/gmt#3104]
```
cat << EOF > pol.txt
> -Z4
NaN NaN
1.2 1.2
1.2 1.8
1.8 1.8
1.8 1.2
1.2 1.2
EOF

makecpt -Chot -T0/5 > t.cpt
gmt psxy -R-2/5/-1/3 -Jx2c pol.txt -B0 -Ct.cpt -W1p -P -K > lixo.ps
```
![lixo](https://user-images.githubusercontent.com/537321/79290859-14bfe980-7ec5-11ea-8dc2-84efaa03c481.png)

**Comment by PaulWessel** [comment]
Beautiful.  I will see what I can learn - drowning in self-inflicted issues at the moment.

**Comment by PaulWessel** [comment]
So this reminds me of what is happening.  You may recall we have this default setting:

**IO_NAN_RECORDS**

> Determines what happens when input records containing NaNs for x or y (and in some cases z) are read. This may happen, for instance, when there is text or other junk present instead of data coordinates, and the conversion to a data value fails and yields a NaN. Choose between skip, which will report how many bad records were skipped, and pass [Default], which will quietly pass these records on to the calling programs. For most programs this will result in output records with NaNs as well, but some will interpret these NaN records to indicate gaps in a series; programs may then use that information to detect segmentation (if applicable).

It is by default pass.  As it does, lines are simply broken up at such points since the NaN NaN acts as a segment header.  For polygons they are not skipped and I supposed ends up as  (0,0) projected coordinates.  If you turn this to skip then you get an error message about a bad point but the remaining points then give the expected plot.  The default pass setting goes back into the dark ages.  However, probably obscure enough to reconsider to be skip.  It seems logical that one should skip crap records, no?

**Comment by joa-quim** [comment]
I suspect it results from the fact that ``gmt_polygon_is_open()`` finds the polygon is open and then ``psxy`` replicates the NaN at end. The cure there seems complicated. The best would be that the IO check if segments start with NaNs and skip them

**Comment by PaulWessel** [comment]
Our postings crossed, please read above.

**Comment by joa-quim** [comment]
Right we reached the same point. But what is a crap record? Binary multi-segment files are separated by NaNs

**Comment by joa-quim** [comment]
And setting ``IO_NAN_RECORDS=skip`` doesn't work from externals. I mean the line to (0,0) is still there.

**Comment by PaulWessel** [comment]
This is what happens when we convert a column and get NaN:

```
if (n_convert == GMT_IS_NAN) {	/* Got a NaN or it failed to decode the string */
	if (GMT->current.setting.io_nan_records || !GMT->current.io.skip_if_NaN[col_pos]) {	/* This field (or all fields) can be NaN so we pass it on */
		GMT->current.io.curr_rec[col_pos] = GMT->session.d_NaN;
		n_ok++;	/* Since NaN is considered an OK result */
	}
	else	/* Cannot have NaN in this column, flag record as bad */
		bad_record = true;
	if (GMT->current.io.skip_if_NaN[col_pos]) set_nan_flag = true;	/* Flag that we found NaN in a column that means we should skip */
	col_no++;		/* Count up number of columns found */
}
```

One solution is to declare that polygons cannot have NaNs and set the default value to skip for polygons in modules that reads polygons.  Or we revert the defaults to skip.  I dont thinks there is much backwards compatibility issues here.

**Comment by PaulWessel** [comment]
The above is for ascii input of course.  For externals it is different.  There is no conversion there so.

**Comment by joa-quim** [comment]
I find it *pacific* that inside segments there should be no NaNs, whatever their geometry is.

**Comment by PaulWessel** [comment]
Setting the default to skip gives me 6 test failures.

**Comment by joa-quim** [comment]
Anyway, wine at dinner told me that this case is not representative of the original one from the shapefiles. Possibilities are
1. Shit (NaNs) in data
2. Bug in ``ogrread``
3. Bug in the Julia function that decodes the output of ``ogrread``

Not a little domain to search

**Comment by joa-quim** [comment]
Found the root of the problem. In ``ogrread`` islands inside polygons are appended to the outer ring and separated by a row of NaNs. I do this to keep track of the paternity of those islands but that is backfiring here.

If one want to paint a polygon with holes, how would we do? Paint the outer and than repaint the holes in white? But what if the background is not white or is another painted polygon?

**Comment by PaulWessel** [comment]
We are handling holes in GMT.  Via shapefiles these are converted in the OGR to have segment headers with -Ph for polygon holes (I think - did not look right now).  So GMT is able to paint polygons with holes correcty (i.e., they become donuts and see-through).  If all you pass is a NaN segment header then we have no way of knowing.

**Comment by joa-quim** [comment]
But what is a segment with a hole? The holes must be passed as different segments, no?
And yes I have other mean than just NaNs to know where hole segments are.

**Comment by PaulWessel** [comment]
Yes, the holes are segments but the headers indicate they are holes.  All the holes in a segment must follow immediately after the parent segment.  Try this:

cat << EOF > pol.txt
>
1.2 1.2
1.2 1.8
1.8 1.8
1.8 1.2
1.2 1.2
> -Ph
1.5 1.5
1.6 1.5
1.6 1.6
1.5 1.6
1.5 1.5
EOF

gmt plot -R-2/5/-1/3 -Jx2c pol.txt -B0 -Gred -W1p -pdf map

**Comment by joa-quim** [comment]
Perfect, I can accommodate that in my ``ogr2GMTdataset()`` julia fun but probably need to add still another member to #3109 to know the total number of segments that I need to allocate in Julia. Need to think.

**Comment by PaulWessel** [comment]
Can this issue be closed, @joa-quim ?

**Comment by joa-quim** [comment]
Well, although it should be a rare occurrence (I fixed the julia case where it occurred), it was not fixed,
